### PR TITLE
CVSL-1906 add approver caseload service to back end

### PR DESCRIPTION
**Pull request by gurps-singh**
This PR is to introduce the approver caseload service to the backend. This includes porting the service that is currently in the frontend to the backend.

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadServiceTest.kt [hunk 1]
```diff
@@ -24,4 +45,151 @@ class ApproverCaseloadServiceTest {
     service.getApprovalNeeded(aListOfPrisonCodes)
     verify(prisonApproverService, times(1)).getLicencesForApproval(listOf("ABC", "DEF"))
   }
+
+  @Test
+  fun `It builds the approval needed caseload`() {
```
Comment (gurps-singh) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadServiceTest.kt:50]: Will be adding more tests in another PR but this matches the unit test on the front end currently.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadService.kt [hunk 2]
```diff
@@ -3,24 +3,141 @@ package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.service.caseload
...
+    val deliusNomisIds = deliusRecords.mapNotNull { it.otherIds.nomsNumber }
+    val nomisRecords = prisonerSearchApiClient.searchPrisonersByNomisIds(deliusNomisIds)
+
+    val prisonerRecord: List<Triple<PrisonerSearchPrisoner, OffenderDetail?, LicenceSummaryApproverView?>> =
+      nomisRecords.map {
+        Triple(it, it.getDeliusRecord(deliusRecords), it.getLicenceSummary(licences))
+      }
+
+    val caseload = prisonerRecord.map { (nomisRecord, deliusRecord, licenceSummary) ->
+      val releaseDate = when (licenceSummary) {
```
Comment (gurps-singh) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadService.kt:51]: Umming and ahhing about this as although this is in the frontend, a licence summary wouldn't be null at this point as it is used to retrieve the `nomisIds` which drive the searches in NOMIS and Delius. Don't think this is needed but tried to match what is on the frontend.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadService.kt [hunk 3]
```diff
@@ -3,24 +3,141 @@ package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.service.caseload
...
+      return null
+    }
+    ProbationPractitioner(
+      staffCode = activeCom.staffDetail.code,
+      name = "${activeCom.staffDetail.forenames} ${activeCom.staffDetail.surname}".convertToTitleCase(),
+    )
   }
 }
+
+data class ApprovalDetails(
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadService.kt:127]: **suggestion:** could these be private classes? Presumably they are only referenced in this file?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadService.kt [hunk 4]
```diff
@@ -3,24 +3,141 @@ package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.service.caseload
...
+fun PrisonerSearchPrisoner.getDeliusRecord(deliusRecords: List<OffenderDetail>): OffenderDetail? {
+  return deliusRecords.find { it.otherIds.nomsNumber == this.prisonerNumber }
+}
+
+fun PrisonerSearchPrisoner.getLicenceSummary(licences: List<LicenceSummaryApproverView>): LicenceSummaryApproverView? {
+  val licenceSummaries = licences.filter { it.nomisId == this.prisonerNumber }
+  return if (licenceSummaries.size == 1) licenceSummaries.first() else licenceSummaries.find { it.licenceStatus != LicenceStatus.ACTIVE }
+}
+
+fun PrisonerSearchPrisoner.selectReleaseDate(): LocalDate? {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadService.kt:103]: **suggestion(non-blocking):** we have this logic on prisoner search prisoner, in lots of places now.. 

Would be good to extract this out and share maybe into its own module in `uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison`

Can see it in:
* LicenceFactory x 2
* EligibilityService

(There's an example in prisonerSearchService which looks similar where we fallback to releaseDate which probably should be CRD too)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadService.kt [hunk 5]
```diff
@@ -3,24 +3,141 @@ package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.service.caseload
...
   }
+}
 
-  companion object {
-    val log: Logger = LoggerFactory.getLogger(this::class.java)
+fun PrisonerSearchPrisoner.getDeliusRecord(deliusRecords: List<OffenderDetail>): OffenderDetail? {
+  return deliusRecords.find { it.otherIds.nomsNumber == this.prisonerNumber }
+}
+
+fun PrisonerSearchPrisoner.getLicenceSummary(licences: List<LicenceSummaryApproverView>): LicenceSummaryApproverView? {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadService.kt:98]: **quibble(non-blocking)**:  I'd probably be tempted to swap the target of the extension function round: 

`licenceSummaries.findFor(prisoner)`

The reason why is that licenceSummaries isn't really an attribute of prisoner search prisoner:
```kt
val prisoners = prisonSearchApi.getPrisoners()
val licence = prisoner.getLicenceSummary()
```
makes it look like the licences are stored in the prisonSearchApi which is potentially a bit confusing.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadService.kt (hunk 2 above)
Comment (andrewrlee) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/ApproverCaseloadService.kt:51]: yes, totally right - this was something in hindsight I thought I should have removed. I was copying over from the existing implementation. 

I think remove it?

**Review by andrewrlee — APPROVED** [review]
LGTM
